gravityflow/gravityflow
# Add build system and icons

## Description

This pr is a bit of a bear. 

![Dance bear dance](https://metro.co.uk/wp-content/uploads/2016/11/pole-dancing-bear.gif)

**Adds:**

- A [Gulp](https://gulpjs.com/) based system that emulates the [PostCSS](https://postcss.org/) build found in Gravity Forms
- An automated icon injection system powered by [IcoMoon](https://icomoon.io/) that also emulates the one found on Gravity Forms
- A modern es6/react capable js build system and custom [webpack](https://webpack.js.org/) config
- Javascript and CSS linting using [eslint](https://eslint.org/) and [stylelint](https://stylelint.io/) that uses the [WordPress standards](https://www.npmjs.com/package/@wordpress/eslint-plugin) 
- Javascript testing and some util tests powered by [Jest](https://jestjs.io/) and [Enzyme](https://enzymejs.github.io/enzyme/)

**High level:**

The PostCSS system added here will generate two css bundles, admin and theme, like GF. Shared styles between the two contexts will live in a shared directory and can be imported to both bundles. I did not take the existing css and convert it to PostCSS in this pr, its a huge task and too risky. Instead, my approach is this: new code will be done in PostCSS and old code when worked on will be migrated, or we can set times to migrate portions and then qa that aspect.

The icon system is not fully documented in the readme, and shouldnt be, I need to know were to document that internal process! The gist of it is that we will now manage all icons in the icomoon app. When you complete adding a new icon, for either admin or theme kit, you save and then get a zip file. Place the zip file in the plugin root and run `gulp icons:admin` or `gulp icons:theme`. The icons will be extracted, postcss files updated, dev kit installed for previewing icons in the dev folder, and the zip will be deleted.

The js system is my proposal for both us and bringing into gf 2.6. It needs a readme entry and also a basic usage guide somewhere. 

It brings in a range of custom utilities I have built over the years as a util system along with a webpack config I have been refining over the last few years, trimmed down to our (future) needs. The actual entry point and module system is very unopinionated, allowing you to encapsulate old school jquery, es6 js classes, react apps, vanilla js in whatever pattern you like. It also enables testing frameworks for the plain js and also any react down the road. 

While the webpack system and some of the build code may be a bit daunting, these systems just work and shouldnt be something anyone has to really do anything with. Babel and webpack will want occasional updating, thats about it. 

One exciting aspect is the ability in this system to hybridize multiple templating approaches with react "micro apps" that can slowly increase in scale as needed, and the helpers i have added for interoperating with the new WordPress internals as needed.

## Testing instructions

Follow the instructions [here](https://github.com/gravityflow/gravityflow/blob/add-build-system-and-icons/readme.md#local-development) to test out local development of with the new system. Run the gulp dev task and test out editing css or js in the src directory. Also run a package build and confirm min files are created. 

## Automated Test Enhancements

This pr does contain a Jest based js testing system ready to go, with 44 tests for various custom utils in place to seed the patterns. It also contains an Enzyme setup ready for testing react components whenever we start to build those down the road.

## Documentation Changes

The readme has been updated, and needs to be enhanced to cover JavaScript patterns and that aspect of the dev flow.

## Checklist:
